src/rebar_packages.erl

```diff
@@ -23,73 +30,97 @@
 -type vsn() :: binary().
 -type package() :: pkg_name() | {pkg_name(), vsn()}.
 
--spec packages(rebar_state:t()) -> ets:tid().
-packages(State) ->
-    catch ets:delete(?PACKAGE_TABLE),
-    case load_and_verify_version(State) of
-        true ->
-            ok;
-        false ->
-            ?DEBUG("Error loading package index.", []),
-            handle_bad_index(State)
-    end.
+format_error({missing_package, Name, Vsn}) ->
+    io_lib:format("Package not found in registry: ~ts-~ts.", [rebar_utils:to_binary(Name), 
+                                                              rebar_utils:to_binary(Vsn)]);
+format_error({missing_package, Pkg}) ->
+    io_lib:format("Package not found in registry: ~p.", [Pkg]).
 
-handle_bad_index(State) ->
-    ?ERROR("Bad packages index. Trying to fix by updating the registry.", []),
-    {ok, State1} = rebar_prv_update:do(State),
-    case load_and_verify_version(State1) of
-        true ->
-            ok;
-        false ->
-            %% Still unable to load after an update, create an empty registry
-            ets:new(?PACKAGE_TABLE, [named_table, public])
+-spec get(hex_core:config(), binary()) -> {ok, map()} | {error, term()}.
+get(Config, Name) ->    
+    case hex_api_package:get(Config, Name) of
+        {ok, {200, _Headers, PkgInfo}} ->
+            {ok, PkgInfo};
+        _ ->
+            {error, blewup}
     end.
 
-close_packages() ->
-    catch ets:delete(?PACKAGE_TABLE).
+-spec get_all_names(rebar_state:t()) -> [binary()].
+get_all_names(State) ->    
+    verify_table(State),
+    lists:usort(ets:select(?PACKAGE_TABLE, [{#package{key={'$1', '_'},
+                                                      _='_'}, 
+                                             [], ['$1']}])).
+
+-spec get_package_versions(binary(), ets:tid(), rebar_state:t()) -> [vsn()].
+get_package_versions(Dep, Table, State) ->
+    ?MODULE:verify_table(State),
+    ets:select(Table, [{#package{key={Dep,'$1'},
+                                 _='_'}, 
+                        [], ['$1']}]).
+
+new_package_table() ->    
+    ets:new(?PACKAGE_TABLE, [named_table, public, ordered_set, {keypos, 2}]),
+    ets:insert(package_index, {?PACKAGE_INDEX_VERSION, package_index_version}).
+
+-spec get_package_deps(binary(), vsn(), rebar_state:t()) -> [map()].
+get_package_deps(Name, Vsn, State) ->
+    try_lookup(?PACKAGE_TABLE, {Name, Vsn}, #package.dependencies, State).
+
+-spec registry_checksum(binary(), vsn(), rebar_state:t()) -> binary().
+registry_checksum(Name, Vsn, State) ->
+    try_lookup(?PACKAGE_TABLE, {Name, Vsn}, #package.checksum, State).
```

Comment:
so this function only looks to see whether _any_ index contains the checksum needed right?

I guess there's a low probability that downloading a package from registry A, containing a checksum that only registry B knows about, would end up accidentally working.

Reply:
No, it is going to have to work on the repo that the package was found in.

Reply:
Alright. I'm curious to see how that's gonna work but this isn't blocking here